Blog posts for 0.8.16. (#171)
ethereum/solidity-blog: _posts/2022-03-16-solidity-0.8.13-release-announcement.md

implementing the functions of the type, ``global`` can only be used
for a user-defined type and only in the same file where the type is defined.
We recommend that the ``using`` directive is always put directly after the definition
of the type (forward references are valid in Solidity).

As an example, let us consider a type that can only be incremented and decremented by one
at a time. Because of this restriction, we can safely drop overflow checks
since it would take more gas than is available in the universe to reach
an overflow. We use the mechanism of an implementation file
so that the names ``plusOne`` and ``minusOne`` do not pollute the namespace.

File ``"restrictedNumber_impl.sol"``:
```
import RestrictedNumber from "./restrictedNumber.sol";
function plusOne(RestrictedNumber x) pure returns (RestrictedNumber)
{
    unchecked {
        return RestrictedNumber.wrap(RestrictedNumber.unwrap(x) + 1);
    }
}
function minusOne(RestrictedNumber x) pure returns (RestrictedNumber)
{
    unchecked {
        return RestrictedNumber.wrap(RestrictedNumber.unwrap(x) - 1);
    }
}
```

File ``"restrictedNumber.sol"``:
```
import "./RestrictedNumber_impl.sol" as RestrictedNumberImpl;

> type RestrictedNumber is int256;

Reviewer: Isn't `uint` more appropriate?

Author: No, because then it can underflow

Reviewer: Hmm, at least for the owner counting case, this would never happen.

using {RestrictedNumberImpl.inc, RestrictedNumberImpl.dec} for RestrictedNumber global;
```

File ``"owned.sol"``:
```
import RestrictedNumber from "./restrictedNumber.sol";
contract Owned {
    RestrictedNumber public ownerCount;

>     mapping(address => bool) public isOwner;

Reviewer: I'd recommend initializing these two variables to make the example somewhat complete.

Author: What do you mean by initialize?

Reviewer: Have a constructor and do
```solidity
isOwner[msg.sender] = true;
ownerCount = ownerCount.plusOne();
```

    function addOwner(address owner) external {

>         require(isOwner[owner]);

Reviewer: ```suggestion
        require(isOwner[msg.sender]);
        isOwner[owner] = true;
```
?

Author: Good one! :)

        // We do not have to add "using for" in the contract
        // to use the ``plusOne`` function.
        ownerCount = ownerCount.plusOne();
    }
    function removeOwner(address owner) external {

>         require(isOwner[owner]);

Reviewer: Same here, right?

---
layout: post
published: true
title: 'Solidity 0.8.13 Release Announcement'
date: '2022-03-16'
author: Solidity Team
category: Releases
---

[Solidity v0.8.13](https://github.com/ethereum/solidity/releases/tag/v0.8.13)
...

## Important Bugs

When ``abi.encodeCall`` was introduced in Solidity 0.8.11,
hex literals (``0x1234``) and string literals (``"abcd"``) were not handled properly.
Please read more about it in the [security alert](https://blog.soliditylang.org/2022/03/16/encodecall-bug/).

## Notable New Features

### Yul IR Pipeline Production Ready

We have been working on Yul as an intermediate language for Solidity for several years now.
Yul in itself has been considered production ready for a while - essentially since
ABI coder V2 was marked "stable", since it makes heavy use of Yul. The other part,
code generation from Solidity to Yul, can now also be considered stable and production ready.

It has been feature complete for several months and since the beginning, we always ran all our
tests through both compiler pipelines. Over the previous weeks, we added several more complex

> real-world contracts to our testing infrastructure and fixed the last bugs in the component.

Author: "last-minute bugs" does not sound right for me.

> real-world contracts to our testing infrastructure and fixed the last bugs in the component.

Author: Pushed a different wording here, please check.

> real-world contracts to our testing infrastructure and fixed the last bugs in the component.

Reviewer 2: Sounds fine to me but maybe "last-minute bugfixes" would be more idiomatic? It's fine now in any case.

is optimized away.

The performance of the new pipeline is not yet always superior to the old one, but it
can do much higher-level optimization across functions, so please try it out and give us feedback!

One important feature of the new pipeline is that it can move stack variables to memory
and thus avoid the "stack too deep" issue. In order to do this safely, the compiler needs to know
that the memory slots it intends to use are not overwritten by inline assembly.
Because of that, inline assembly should only use memory that has been previously allocated
either by high-level Solidity code or by reading from the free memory pointer at ``0x40``.
At the very beginning of the code, the compiler will set the free memory pointer to a position
after the area reserved for the stack variables that were moved to memory.
Since it is difficult to impossible to check by the compiler that inline assembly only accesses memory in a safe way,
we added a syntactic mechanism for the developer to assert this:

```
assembly ("memory-safe") { ... }
```

It can lead to undefined behaviour if you mark an inline assembly block as memory safe if it is

> not actually memory safe, because it can lead to conflicts between stack variables and memory.

Reviewer 3: Maybe we should link to the docs section with details about what's memory safe?

The module that can move stack variables to memory is only activated if all inline assembly
blocks either do not access memory at all or are marked as being memory safe.

If you are e.g. writing a code library that should be used across compiler versions,
you can also use a special comment syntax that has the same effect:
``/// @solidity memory-safe-assembly``

We advise against using this syntax, though, as we also advise against using the same

> code base across breaking compiler versions.

Reviewer 3: I don't really see the connection to "using the same code base across breaking compiler versions"...
It's neither a breaking release, nor does this have any danger of silent semantics changes one would have in a breaking release...
What we do advice is to always use the latest compiler version - that has more connection to this, doesn't it?

Author: My point is, if you are using the comment with `pragma solidity ^0.8.0`, you can just as well use the proper way and require 0.8.13. The main reason for using the comment is that you are also supporting 0.7.x or older.

Reviewer 3: Does that make sense :-)? Anyone supporting both 0.7.x and 0.8.x is rather unlikely due to checked arithmetic alone, isn't it? I was under the impression that the comment was *exactly* meant for supporting ``^0.8.0`` in a library (based on arguments like https://forum.soliditylang.org/t/thoughts-on-override-a-b-syntax/767/14)... if we say anything using ``^0.8.0`` should just jump to 0.8.13, I don't really see the comment version to be of much use after all :-)...

Author: Oh yeah that is actually a good point. I mean you could have no arithmetics at all...

Reviewer 2: Looking at the external contracts I had to convert to 0.8.x I remember there being some unaffected by safe math so I don't think having such a library is that unlikely.

to design proper reusable code libraries.

In a future release, we plan to also add operators to user-defined types
so that things like fixed point type libraries are relly easy to use
and feel like compiler built-in types.

If you use ``using {f, g, h} for Type;``, then the compiler already
verifies that the given functions can receive a ``Type`` value as first
argument.

You can add the word ``global`` at the end to make the effect available
everywhere. Since the idea is that you are defining a type and
implementing the functions of the type, ``global`` can only be used
for a user-defined type and only in the same file where the type is defined.
We recommend that the ``using`` directive is always put directly after the definition
of the type (forward references are valid in Solidity).

As an example, let us consider a type that can only be incremented and decremented by one
at a time. Because of this restriction, we can safely drop overflow checks
since it would take more gas than is available in the universe to reach
an overflow. We use the mechanism of an implementation file
so that the names ``plusOne`` and ``minusOne`` do not pollute the namespace.

> File ``"restrictedNumber_impl.sol"``:

Reviewer 2: ```suggestion
File ``restrictedNumber_impl.sol``:
```

```

> import RestrictedNumber from "./restrictedNumber.sol";
> function plusOne(RestrictedNumber x) pure returns (RestrictedNumber)
> {
>     unchecked {
>         return RestrictedNumber.wrap(RestrictedNumber.unwrap(x) + 1);
>     }
> }
> function minusOne(RestrictedNumber x) pure returns (RestrictedNumber)

Reviewer 2: I'd add some spacing for readability. It's not like we're that constrained here.
```suggestion
import RestrictedNumber from "./restrictedNumber.sol";

function plusOne(RestrictedNumber x) pure returns (RestrictedNumber)
{
    unchecked {
        return RestrictedNumber.wrap(RestrictedNumber.unwrap(x) + 1);
    }
}

function minusOne(RestrictedNumber x) pure returns (RestrictedNumber)
```

{
    unchecked {
        return RestrictedNumber.wrap(RestrictedNumber.unwrap(x) - 1);
    }
}
```

> File ``"restrictedNumber.sol"``:

Reviewer 2: ```suggestion
File ``"RestrictedNumber.sol"``:
```

```

> import "./RestrictedNumber_impl.sol" as RestrictedNumberImpl;

Reviewer 2: Why not use the same convention for file name and module name?
```suggestion
import "./RestrictedNumberImpl.sol" as RestrictedNumberImpl;
```

type RestrictedNumber is int256;

> using {RestrictedNumberImpl.inc, RestrictedNumberImpl.dec} for RestrictedNumber global;

Reviewer 2: Wouldn't it make the example simpler to define these in the same file together with the type?

Author: Probably. I wanted to show how you can hide stuff, but maybe that's meant for another blog post.

/// Apart from default-initializing, inline assembly and passing the value
/// through a transaction, this is the only way to create a value of that type.
/// If we don't use any of the other methods, dropping the overflow check
/// is safe.
function createRestrictedNumber(int256 value) pure returns (RestrictedNumber)
{
    // Ensure that the number is "small".
    // Larger constants like 2**200 would also work.
    require(value <= 100 && -value <= 100);
    return RestrictedNumber.wrap(value);
}
```

File ``"owned.sol"``:
```
import RestrictedNumber from "./restrictedNumber.sol";
contract Owned {
    RestrictedNumber public ownerCount;

>     mapping(address => bool) public isOwner;
>     constructor() {
>         _addOwner(msg.sender);
>     }
>     function addOwner(address owner) external {
>         require(isOwner[msg.sender]);
>         _addOwner(owner);
>     }
>     function removeOwner(address owner) external {
>         require(isOwner[msg.sender]);
>         require(isOwner[owner]);
>         ownerCount = ownerCount.minusOne();
>         isOwner[owner] = false;
>     }
>     function _addOwner(address owner) internal {

Reviewer 2: This could use some spacing too:
```suggestion
    mapping(address => bool) public isOwner;

    constructor() {
        _addOwner(msg.sender);
    }

    function addOwner(address owner) external {
        require(isOwner[msg.sender]);
        _addOwner(owner);
    }

    function removeOwner(address owner) external {
        require(isOwner[msg.sender]);
        require(isOwner[owner]);
        ownerCount = ownerCount.minusOne();
        isOwner[owner] = false;
    }

    function _addOwner(address owner) internal {
```

        require(!isOwner[owner]);

>         // We do not have to add "using for" in the contract

Reviewer 2: ```suggestion
        // Because of `global`, we do not have to add `using for` in the contract
```

        // to use the ``plusOne`` function.
        ownerCount = ownerCount.plusOne();
        isOwner[owner] = true;
    }
}
```

> ### Goto Definition for LSP

Reviewer 2: ```suggestion
### "Go to Definition" for LSP
```

The Solidity Language Server has been further improved and received

> one new feature, going to the definition of the underlying symbol.

Reviewer 2: This sounds a bit awkward. How about this?
```suggestion
one new feature: definition lookup of the underlying symbol.
```

This works on every symbol, such as local variables, return variables,
type names, but also on import statements.

We consider the official Solidity Language Server still as a preview,

> that is, while being functional, there will be much more coming.

Reviewer 2: ```suggestion
that is, while it is already functional, there will be much more coming.
```

The module that can move stack variables to memory is only activated if all inline assembly
blocks either do not access memory at all or are marked as being memory safe.

If you are e.g. writing a code library that is meant to be used across compiler versions,
you can also use a special comment syntax that has the same effect:
/// @solidity memory-safe-assembly
assembly { ... }

We advise against using this syntax, though, as we also advise against using the same
code base across breaking compiler versions.

### `using for` at File Level with Global Binding

This release extends the ``using for`` directive considerably. It can now be used
at file level, it is possible to use it with free functions and
it has the option to make its effect global. We believe that in combination
with the recently released user-defined value types, this makes it possible
to design proper reusable code libraries.

In a future release, we plan to also add operators to user-defined types
so that things like fixed-point type libraries are really easy to use
and feel like compiler built-in types.

If you use ``using {f, g, h} for Type;``, then the compiler

> verifies that the given functions can receive a ``Type`` value as first

Author: I'll rephrase that and remove the type checking comment, it's really not that relevant.

is primarily meant to be a straightforward transformation of the high-level Solidity code
and it has many additional checks compared to the old compiler pipeline.
This makes it more
modular and easier to audit while the new Yul-based optimizer is powerful enough to remove redundancies
from the optimized version.

The performance of the new pipeline is not yet always superior to the old one, but it
can do much higher-level optimization across functions, so please try it out and give us feedback!

Finally, in some edge cases, code compiled via the new compiler pipeline
behaves different than code compiled via the old compiler pipeline.
The most prominent is the initialization order of state variables that depend
on the result of constructors in other contracts.
Others are related to weird uses of modifiers, dependence of evaluation order inside
statements or dirty data. You can find a list of all the changes
in our [documentation](https://docs.soliditylang.org/en/develop/ir-breaking-changes.html).

### Memory-Safe Assembly / Stack Too Deep

One important feature of the new compiler pipeline is that it can move stack variables to memory

> and thus avoid the "stack too deep" issue. In order to do this safely, the compiler needs to know

Reviewer 3: ```suggestion
and thus avoid the "stack too deep" issue in a lot of cases. In order to do this safely, the compiler needs to know
```
to be a bit more conversative :-)... we probably don't need to mention that it won't happen in recursive functions, but we should mention that it's only done by the optimizer.

that the memory slots it intends to use are not used by inline assembly.

Inline assembly is considered memory-safe if it only uses memory that has been previously allocated
either by high-level Solidity code or by reading from the free memory pointer at ``0x40``.
At the very beginning of the code, the compiler sets the free memory pointer to a position
after the area reserved for the stack variables that are moved to memory.

Since it is difficult to impossible for the compiler to check that inline assembly only accesses memory in a safe way,
we added a syntactic mechanism for the developer to assert this:

```
assembly ("memory-safe") { ... }
```

Marking an unsafe block as memory-safe can lead to undefined behavior
due to conflicts between stack variables and memory, the compiler will
not check that marked blocks are actually memory safe.
Please see the [documentation](https://docs.soliditylang.org/en/v0.8.13/assembly.html#memory-safety)
for more information about when an inline assembyl block is memory safe.

> By default inline assembly blocks are not considered memory-safe and

Reviewer 3: Not strictly true for inline assembly blocks without memory accesses and without references to solidity variables in memory - but not sure we need to mention that, so also fine.

Author: I'll add it.

marking it "non-experimental" earlier was that it still had problems with limited stack space,
which we were able to fix now.

You can enable the new pipeline using the ``--via-ir`` commandline switch or by setting
``settings.viaIR`` to ``true`` in Standard JSON input (i.e. at the same level where
you would have the ``optimizer`` key).

It is not recommended to use the Yul compiler pipeline without the optimizer because the unoptimized output
is primarily meant to be a straightforward transformation of the high-level Solidity code
and it has many additional checks compared to the old compiler pipeline.
This makes it more
modular and easier to audit while the new Yul-based optimizer is powerful enough to remove redundancies
from the optimized version. Without the optimizer, you are also more likely to run into "stack too deep"
issues which can often be avoided by the optimizer as explained below.

The performance of the new pipeline is not yet always superior to the old one, but it
can do much higher-level optimization across functions, so please try it out and give us feedback!

Finally, in some edge cases, code compiled via the new compiler pipeline

> behaves different than code compiled via the old compiler pipeline.

Reviewer 2: ```suggestion
behaves differently than code compiled via the old compiler pipeline.
```

### Memory-Safe Assembly / Stack Too Deep

One important feature of the new compiler pipeline is that the Yul optimizer can move stack variables to memory
and thus avoid the "stack too deep" issue in a lot of cases. In order to do this safely, the compiler
needs to know that the memory slots it intends to use are not used by inline assembly.

Inline assembly is considered memory-safe if it only uses memory that has been previously allocated
either by high-level Solidity code or by reading from the free memory pointer at ``0x40``.
At the very beginning of the code, the compiler sets the free memory pointer to a position
after the area reserved for the stack variables that are moved to memory.

The compiler considers inline assembly blocks as memory safe only for
very simple blocks that do not have any opcodes that access memory and also do not access
Solidity variables related to memory reference types.
For all other blocks, it is difficult to impossible to determine this automatically in a safe way
and thus we added a syntactic mechanism for the developer to assert it:

```
assembly ("memory-safe") { ... }
```

> Marking an unsafe block as memory-safe can lead to undefined behavior
> due to conflicts between stack variables and memory, the compiler will
> not check that marked blocks are actually memory safe.

Reviewer 2: Run-on sentences.
```suggestion
Marking an unsafe block as memory-safe can lead to undefined behavior
due to conflicts between stack variables and memory. The compiler will
not check that marked blocks are actually memory safe.
```

Please see the [documentation](https://docs.soliditylang.org/en/v0.8.13/assembly.html#memory-safety)

> for more information about when an inline assembyl block is memory safe.

Reviewer 2: ```suggestion
for more information about when an inline assembly block is memory safe.
```

Inline assembly blocks that access memory are not considered memory-safe by default and
even a single such block disables the stack-to-memory mechanism for the whole contract.

> Assuming otherwise would be dangerous and due to the risk of stack variables moved to memory being overwritten.

Reviewer 2: ```suggestion
Assuming otherwise would be dangerous due to the risk of stack variables moved to memory being overwritten.
```

If you are e.g. writing a code library that is meant to be used across compiler versions,
you can also use a special comment syntax that has the same effect:

> /// @solidity memory-safe-assembly
> assembly { ... }

Reviewer 2: This is missing a code block.
````suggestion
```
/// @solidity memory-safe-assembly
assembly { ... }
```
````

Now that I think of it, I probably used the triple backtick inside a `suggestion` block in my previous suggestion so this might not have been visible.
